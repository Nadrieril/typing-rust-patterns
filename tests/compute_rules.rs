use std::fmt::Write;

use typing_rust_patterns::*;

#[test]
fn compute_rules() -> anyhow::Result<()> {
    let arenas = &Arenas::default();
    for &(name, options, _) in RuleOptions::KNOWN_OPTION_BUNDLES {
        let ctx = TypingCtx { arenas, options };

        let mut typing_rules = typing_rust_patterns::compute_rules(ctx);
        typing_rules.sort_by_key(|rule| rule.name);

        let mut rules_str = String::new();
        for rule in typing_rules {
            let _ = writeln!(&mut rules_str, "{rule}\n");
        }

        insta::with_settings!({
            snapshot_suffix => name,
            info => &name,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(rules_str);
        });
    }

    Ok(())
}
