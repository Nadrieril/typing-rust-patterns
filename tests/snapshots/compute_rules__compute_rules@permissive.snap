---
source: tests/compute_rules.rs
info: permissive
---
p0 @ q.0: T0, p1 @ q.1: T1
-------------------------- "Constructor"
[p0, p1] @ q: [T0, T1]

p0 @ &(*q).0: &T0, p1 @ &(*q).1: &T1
------------------------------------ "ConstructorRef"
[p0, p1] @ q: &[T0, T1]

p0 @ &mut (*q).0: &mut T0, p1 @ &mut (*q).1: &mut T1
---------------------------------------------------- "ConstructorRef"
[p0, p1] @ q: &mut [T0, T1]

[p0, p1] @ *q: &T
----------------- "ConstructorMultiRef"
[p0, p1] @ q: &&T

[p0, p1] @ &**q: &T
--------------------- "ConstructorMultiRef"
[p0, p1] @ q: &&mut T

[p0, p1] @ *q: &T
--------------------- "ConstructorMultiRef"
[p0, p1] @ q: &mut &T

[p0, p1] @ &mut **q: &mut T
--------------------------- "ConstructorMultiRef"
[p0, p1] @ q: &mut &mut T

p @ *q: T
---------- "Deref(EatOuter)"
&p @ q: &T

p @ *q: T
------------------ "Deref(EatOuter)"
&mut p @ q: &mut T

&p @ &*q: &T
-------------------------------------- "DerefMutWithShared(EatOuter)"
&p @ q: &mut T, binding_mode(q) = move

&p @ q: &T
----------------------------------------- "DerefMutWithShared(EatOuter)"
&p @ q: &mut T, binding_mode(q) = ref mut

x @ &q: &T
------------ "BindingBorrow"
ref x @ q: T

x @ &mut q: &mut T
------------------ "BindingBorrow"
ref mut x @ q: T

let x: T = q
------------ "Binding"
x @ q: T

let mut x: T = q
---------------- "Binding"
mut x @ q: T

p @ p: T
-------------- "ExprSimplification"
p @ *&mut p: T
